File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
answer = 0

def dfs(begin, target, words, cnt):
    global answer

    if begin == target:
        answer = cnt

    for word in words:
        dif = 0
        for i in range(len(word)):
            if begin[i] != word[i]:
                dif += 1

        if dif == 1:
            new = words[:]
            new.remove(word)
            dfs(word, target, new, cnt+1)
            
def solution(begin, target, words):
    dfs(begin, target, words, 0)
    return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    int answer = 0;
    
    void dfs(String begin, String target, String[] words, int count) {
        if (begin.equals(target)) {
            answer = count;
            return;
        }
        
        for (String word: words) {            
            int dif = 0;
            for (int i = 0; i < word.length(); i++) {
                if (begin.charAt(i) != word.charAt(i)) {
                    dif += 1;
                }
            }
                        
            if (dif == 1) {
                List<String> wordList = new ArrayList<>();
                for (String w : words) {
                    if (!w.equals(word)) {
                        wordList.add(w);
                    }
                }
                String[] copy = wordList.toArray(new String[0]);
                dfs(word, target, copy, count + 1);
            }
        }
    }
    
    public int solution(String begin, String target, String[] words) {
        dfs(begin, target, words, 0);
        return answer;
    }
}